core/src/main/scala/tofu/Context.scala

```diff
@@ -66,98 +122,260 @@ object Context {
+    */
   def local[A](fa: F[A])(project: Ctx => Ctx): F[A]
 
+  /** Allows to focus [[Local]] on inner parts of its context with lens.
+    *
+    * @param contains lens that can extract from `Ctx` or set some value of type `A`
+    */
   def subcontext[A](contains: Ctx Contains A): WithLocal[F, A] = new LocalContainsInstance[F, Ctx, A](this, contains)
 }
 
 object Local {
   def apply[F[_]](implicit ctx: Local[F]): HasLocal[F, ctx.Ctx] = ctx
+
   type Aux[F[_], C] = HasLocal[F, C]
 }
 
-/** lightweight version of ApplicativeLocal
-  * allows to modify value of `C` locally for some subprocess
-  */
+/** Synonym for [[Local]] with explicit C as Ctx for better type inference */
 trait WithLocal[F[_], C] extends Local[F] with WithContext[F, C]
 
+/** Companion object for [[WithLocal]] */
 object WithLocal {
   def apply[F[_], C](implicit ctx: WithLocal[F, C]): WithLocal[F, C] = ctx
 }
 
+/** Allows to evaluate contextual computation with some context
+  *
+  * The main use case for it is to obtain some context `Lower[Ctx]`,
+  * and then to run another contextual computation F[A] that can use `F[Ctx]` inside.
+  *
+  * @tparam F context-aware effect e.g.`ReaderT[Lower, Ctx, *]`
+  */
 trait Provide[F[_]] extends ContextBase {
   type Ctx
+
+  /** Result effect type without any notion of context */
   type Lower[A]
 
+  /** Runs `fa` providing `ctx` to it.
+    *
+    * One can treat F as function of type `Ctx => Lower[A]` so this method applies it to `ctx`.
+    *
+    * @example
+    * Example of usage is to hide sensitive information across a part of service {{{
+    *   import tofu.syntax.context._
+    *
+    *   case class UserContext(
+    *       id: String,
+    *       phoneNumber: String //very sensitive
+    *   )
+    *   def hidePartOfPhoneNumber: String => String = ???
+    *
+    *   def contextualLogInfo[F[_]](message: String)(implicit hasUserContext: UserContext In F): F[Unit] =
+    *     ??? //logs both message AND UserContext
+    *
+    *   def program[F[_]: WithRun[*[_], G[_], UserContext], G[_]: Monad] = for {
+    *     user   <- obtainUserSomehow[G]
+    *     logUser = contextualLogInfo[F](s"Successfully obtained user")
+    *     _      <- runContext[F](logUser)(user.toContext) //G[Unit]
+    *   } yield ()
+    * }}}
+    * @param fa Contextual computation
+    * @return Result of running fa and providing it context
+    */
   def runContext[A](fa: F[A])(ctx: Ctx): Lower[A]
+
+  /** Same as [[runContext]] but higher-kinded.
+    *
+    * @example {{{
+    *   trait ProcessHandler[G[_]] {
+    *     def mapK[M[_]](fk: G ~> M): ProcessHandler[M] = ???
+    *     //...other methods
+    *   }
+    *   type WithMyContext[F[_], A] = ReaderT[F, MyCtx, A]
+    *   val contextualProcessHandler: ProcessHandler[IO WithMyContext *] = ???
+    *
+    *   val processHandler: ProcessHandler[IO] =
+    *     processHandler.mapK(
+    *       WithProvide[IO WithMyContext *, IO, MyCtx].runContextK
+    *     )
```

Comment:
@FunFunFine, maybe we can put simpler examples here without all these vogue modules and `???`? I think such examples better suit for an article providing some use case context along with the code.

Reply:
Same for other examples